infra/src/main/kotlin/fi/fta/geoviite/infra/publication/Publication.kt

```diff
@@ -364,16 +366,21 @@ data class LocationTrackChanges(
     val trackNumberId: Change<IntId<TrackLayoutTrackNumber>>,
 )
 
+// Todo: Consider making TrackLayoutSwitch use this for trapPoint as well
+enum class TrapPoint {
+    Yes, No, Unknown
+}
```

Comment:
Tämä tuli uutena. Kommentti oikeastaan kertookin, että tämä olisi vähän niinkuin samalla keskustelun aloituspiste. Julkaisuloki sivuuttaa tarkoituksella `null`-arvot, joten vähintään siellä nämä tarvii muuttaa johonkin muuhun muotoon... Mutta kun kyseessä on kerran vähän niinkuin tri-state-enum, niin miksi ei vaan käsiteltäisi sitä kaikkialla sellaisena? Kantaanhan sen voi edelleen halutessaan tallettaa silti booleanina

Reply:
Olen itse aina ollut vähän unknown enum-arvoja vastaan, joskaan en mitenkään absoluuttisesti. Joskus tuo on helpompaa mutta lähtökohtaisesti pidän luontevampana tulkita enumit niin että ne kuvaavat vain oikeat arvon vaihtoehdot ja puuttuva arvo... puuttuu. Koitanpa vähän jäsentää miksi.

Samaa unknown-arvon logiikkaa voisi venyttää myös esimerkiksi vaihdetyyppiin - onko UNKNOWN myös vaihdetyyppi? Minusta ei, vaan siellä missä vaihdetyyppi voi puuttuua, pitää olla SwitchType? ja siellä missä se on varmasti, voi olla SwitchType. Jos tuntematon arvo leivotaan osaksi enumia, se on kaikkialla mahdollinen ja täytyy kaikkialla käsitellä. Funktio ei voi esim rajapinnassaan sanoa että tätä ei pitäisi tyhjällä arvolla edes kutsua (kuten nullin kanssa voi). Sama sinällään pätee tähänkin - olisi ihan validi business-vaatimus sanoa että kaikista tallennetuista vaihteista pitää tietää onko se trap vai ei, mutta se voi puuttua jossain muussa kontekstissa (esim save-olio). Tällöin enumin sisältämä unknown valuu kaikkialle missä enumia käytetään, vaikka se ei kaikkialla olisi laillinen.

Toinen tilanne on esim elvis-operaattorit. Jos funktio ottaa sisäänsä Switch? ja haluaa passata siitä toiselle funktiolle tuon vaihteen tyypin, välittääkö se `switch?.type ?: UNKNOWN`? Tuossa syntyy helposti toinen tyyppi tuntemattomalle tiedolle (nullable enum arvo vaikka enum itse sisältää jo tuntemattoman).

Reply:
... ja vasta-argumenttina itselleni, esim Javassa tämä on selvästi järkevämpää koska nullit ovat niin iso riesa ja monessa koodipohjassa koitetaan rajoittaa niitä erilaisilla konventioilla. UNKNOWN-arvolla voi monessa paikassa tehdä sen ilman tarvetta Optional:lle yms. Sielläkin itse suosisin kyllä mielummin Optionalia ja erityisesti Kotlinissa null on tyyppi itsessään joten en näe tarvetta piilottaa sitä enumin sisään.